test/unit/org/apache/cassandra/metrics/NativeMemoryMetricsTest.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.metrics;
+
+import java.lang.management.BufferPoolMXBean;
+import java.lang.management.ManagementFactory;
+import java.nio.ByteBuffer;
+
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.utils.memory.MemoryUtil;
+
+import static org.junit.Assert.assertEquals;
+
+public class NativeMemoryMetricsTest
+{
+    private final static Logger logger = LoggerFactory.getLogger(NativeMemoryMetricsTest.class);
+    private static NativeMemoryMetrics nativeMemoryMetrics;
+    private static BufferPoolMXBean directBufferPool;
+
+    @BeforeClass
+    public static void setupClass()
+    {
+        nativeMemoryMetrics = NativeMemoryMetrics.instance;
+
+        directBufferPool = ManagementFactory.getPlatformMXBeans(BufferPoolMXBean.class)
+                                            .stream()
+                                            .filter(bpMBean -> bpMBean.getName().equals("direct"))
+                                            .findFirst()
+                                            .orElse(null);
+
+        // touch every field under test that may result in additional allocations (see MemoryUtil static init)
+        nativeMemoryMetrics.totalMemory();
+        nativeMemoryMetrics.totalNioDirectMemory();
+        nativeMemoryMetrics.usedNioDirectMemory();
+        nativeMemoryMetrics.nioDirectBufferCount();
+    }
+
+    @Test
+    public void testNioDirectMemory()
+    {
+        long totalNioDirectMemory = nativeMemoryMetrics.totalNioDirectMemory();
+        long usedNioDirectMemory = nativeMemoryMetrics.usedNioDirectMemory();
+        long nioDirectBufferCount = nativeMemoryMetrics.nioDirectBufferCount();
+
+        logger.debug("Total Nio Memory: {}, Reserved Nio Memory: {}, Num Nio buffers: {}",
+                     totalNioDirectMemory, usedNioDirectMemory, nioDirectBufferCount);
+
+        assertEquals("Total and reserved nio memory should be equal since -Dsun.nio.PageAlignDirectMemory=true should not be set",
```

Comment:
Shall we also check that `sun.nio.PageAlignDirectMemory` is false?

Reply:
done